File: scilab/modules/linear_algebra/src/fortran/intdgeev.f

      subroutine intdgeev(fname)

c     [V,D]=spec(A)
c     d = spec(A)

      include 'stack.h'
      logical getrhsvar,createvar
      logical checklhs,checkrhs
      integer vfinite

      character fname*(*)
      character JOBVL, JOBVR
      double precision DZERO
      parameter ( DZERO=0.0D0 )
      complex*16 ZERO
      parameter ( ZERO=(0.0D0,0.0D0) )
      INTRINSIC  DCMPLX

      minrhs=1
      maxrhs=1
      minlhs=1
      maxlhs=2
c     
      if(.not.checkrhs(fname,minrhs,maxrhs)) return
      if(.not.checklhs(fname,minlhs,maxlhs)) return 

      if(.not.getrhsvar(1,'d', M, N, lA)) return
      if(M.ne.N) then
         err=1
         call error(20)
         return
      endif
      if(N.eq.0) then
         if(lhs.eq.1) then
            if(.not.createvar(2,'z', 0, 0, lD)) return
            lhsvar(1) = 2
            return
         else if(lhs.eq.2) then
            if(.not.createvar(2,'z', 0, 0, lD)) return
            if(.not.createvar(3,'d', 0, 0, lV)) return
            lhsvar(1) = 2
            lhsvar(2) = 3
            return
         endif
      elseif(N.eq.-1) then
         err=1
         call error(271)
         return
      endif 
      if(vfinite(M*N,stk(lA)).eq.0) then
         err=1
         call error(264)
         return
      endif  
      
      if(lhs.eq.1) then
         if(.not.createvar(2,'z', N, 1, lD)) return
         k = 3              
      else if(lhs.eq.2) then
         if(.not.createvar(2,'z', N, N, lD)) return
         if(.not.createvar(3,'d', N, N, lVR)) return
         if(.not.createvar(4,'z', N, N, lV)) return
         k = 5
      endif
      if(.not.createvar(k,'d', N, 1, lWR)) return
      if(.not.createvar(k+1,'d', N, 1, lWI)) return
      LWORKMIN = 3*N
      if(lhs.gt.1) LWORKMIN = 4*N
      LWORK=maxvol(k+2,'d')
      if(LWORK.le.LWORKMIN) then
         err=(LWORK-LWORKMIN)
         call error(17)
         return
      endif
      if(.not.createvar(k+2,'d',1,LWORK,lDWORK)) return
      JOBVL = 'N'
      if(lhs.eq.1) then
         JOBVR = 'N'
      else
         JOBVR = 'V'
      endif 
      call DGEEV( JOBVL, JOBVR, N, stk(lA), N, stk(lWR), stk(lWI),
     $     stk(lDWORK), N, stk(lVR), N, stk(lDWORK), LWORK, INFO )
c     SUBROUTINE DGEEV( JOBVL, JOBVR, N, A, LDA, WR, WI, VL, LDVL, VR,
c     LDVR, 
c     $     WORK, INFO )
      if(info.ne.0) then
         call error(24)
         return
      endif

      if(lhs.eq.1) then
         call ZLASET( 'F', N, 1, ZERO, ZERO, zstk(lD), N ) 
c     SUBROUTINE ZLASET( UPLO, M, N, ALPHA, BETA, A, LDA )

         do 5 i = 1, N
            zstk(lD+i-1) = dcmplx(stk(lWR+i-1),stk(lWI+i-1))
 5       continue     
      else 
         call ZLASET( 'F', N, N, ZERO, ZERO, zstk(lD), N ) 
c     SUBROUTINE ZLASET( UPLO, M, N, ALPHA, BETA, A, LDA )

         do 10 i = 1, N
            ii = i+(i-1)*N
            zstk(lD+ii-1) = dcmplx(stk(lWR+i-1),stk(lWI+i-1))
 10      continue   
         
         j = 0
 20      j = j+1
         if(stk(lWI+j-1).eq.DZERO) then
            do 30 i = 1, N
               ij = i+(j-1)*N
               zstk(lV+ij-1) = stk(lVR+ij-1)
 30         continue
         else   
            do 40 i = 1, N
               ij = i+(j-1)*N
               ij1 = i+j*N
               zstk(lV+ij-1) = dcmplx(stk(lVR+ij-1),stk(lVR+ij1-1))
               zstk(lV+ij1-1) = dcmplx(stk(lVR+ij-1),-stk(lVR+ij1-1)) 
 40         continue
            j = j+1
         endif
         if(j.lt.N) go to 20
      endif
      
      if(lhs.eq.1) then
         lhsvar(1) = 2
      else 
         lhsvar(1)=4
         lhsvar(2)=2
      endif
c     
      end

File: scilab/modules/linear_algebra/src/c/intdgeev.c
/*
 * Scilab ( http://www.scilab.org/ ) - This file is part of Scilab
 * Copyright (C) 2008 - INRIA - Michaël Baudin
 * 
 * This file must be used under the terms of the CeCILL.
 * This source file is licensed as described in the file COPYING, which
 * you should have received as part of this distribution.  The terms
 * are also available at    
 * http://www.cecill.info/licences/Licence_CeCILL_V2-en.txt
 *
 */
#include "stack-c.h" 
#include "elementary_functions.h" 
#include "do_error_number.h"
#include "MALLOC.h"
#include "assembleEigenvectors.h"
extern int C2F(dsyev)();
extern int C2F(dlaset)();
extern int C2F(dcopy)();
extern int C2F(dgeev)();


//
// intdgeev --
//   Interface to LAPACK's DGEEV
//   Computes the eigenvalues and, if required, the eigenvectors of a un-symetric real matrix.
//   Possible uses :
//   * With 1 LHS :
//       eigenvalues=spec(A)
//     where
//       A : un-symetric, square real matrix of size NxN
//       eigenvalues : matrix of size Nx1 with right eigenvectors, type complex
//   * With 2 LHS :
//       [eigenvectors,eigenvalues]=spec(A)
//     where 
//       eigenvalues : matrix of size NxN with eigenvalues as diagonal terms
//       eigenvectors : matrix of size NxN with right eigenvectors, type complex
//
int intdgeev(char *fname)
{	
	int totalsize;
	int iRows = 0;
	int iCols = 0;
	int ONE = 1;
	int ZERO = 0;
	int iWorkSize;

	int INFO;

	char JOBVR;
	char JOBVL;

	double dlbZero = 0;

	double *pdblWork = NULL;
	double *pdblRealData	= NULL;
	double *pdblFinalEigenvaluesReal = NULL; //SCILAB return Var
	double *pdblFinalEigenvaluesImg = NULL; //SCILAB return Var
	double *pdblFinalEigenvectorsReal = NULL; //SCILAB return Var
	double *pdblFinalEigenvectorsImg = NULL; //SCILAB return Var
	double *pdblEigenvalReal	= NULL; //return by LAPACK
	double *pdblEigenvalImg	= NULL; //return by LAPACK
	double *pdblRightvectors	= NULL; //return by LAPACK
	double *pdblLeftvectors	= NULL; //return by LAPACK

	CheckRhs(1,1) ;
	CheckLhs(1,2) ;

	GetRhsVarMatrixDouble(1, &iRows, &iCols, &pdblRealData);
	totalsize = iRows * iCols;

	if (iRows!=iCols)
	{
		SciError(20);
		return 0;
	}

	if (iCols==0)
	{
		if (Lhs==1)
		{
			int lD;
			int ZERO = 0;
			CreateVar(2,MATRIX_OF_COMPLEX_DATATYPE,&ZERO,&ZERO,&lD);
			LhsVar(1) = 2;
			return 0;
		}
		else if (Lhs==2)
		{
			int lD;
			int ZERO = 0;
			int lV;
			CreateVar(2,MATRIX_OF_COMPLEX_DATATYPE,&ZERO,&ZERO,&lD);
			CreateVar(3,MATRIX_OF_COMPLEX_DATATYPE,&ZERO,&ZERO,&lV);

			LhsVar(1) = 2;
			LhsVar(2) = 3;

			return 0;
		}
	}

	if (C2F(vfinite)(&totalsize,pdblRealData)==0)
	{
		SciError(264);
		return 0;
	}
	if (Lhs==1)
	{
		iAllocMatrixOfDoubleComplex(2, iCols, ONE, &pdblFinalEigenvaluesReal, &pdblFinalEigenvaluesImg);
	}
	else
	{
		iAllocMatrixOfDoubleComplex(2, iCols, iCols, &pdblFinalEigenvaluesReal, &pdblFinalEigenvaluesImg);
		iAllocMatrixOfDoubleComplex(3, iCols, iCols, &pdblFinalEigenvectorsReal, &pdblFinalEigenvectorsImg);
		pdblRightvectors = (double*)MALLOC(totalsize * sizeof(double));
	}
	pdblEigenvalReal = (double*)MALLOC(iCols * sizeof(double));
	pdblEigenvalImg = (double*)MALLOC(iCols * sizeof(double));

	if (Lhs==2)
	{
		iWorkSize = 4*iCols;
	}
	else
	{
		iWorkSize = 3*iCols;
	}
	pdblWork = (double*)MALLOC(iWorkSize * sizeof(double));


    JOBVL = 'N'; // Never compute left eigenvectors
	if (Lhs==1)
	{
		JOBVR = 'N'; // Do not compute right eigenvectors
	}
	else
	{
		JOBVR = 'V'; // Compute right eigenvectors.
	}
	C2F(dgeev)( &JOBVL, &JOBVR, &iCols, pdblRealData, &iCols, pdblEigenvalReal, pdblEigenvalImg,
        pdblLeftvectors, &iCols, pdblRightvectors, &iCols, pdblWork, &iWorkSize, &INFO );
//     SUBROUTINE DGEEV( JOBVL, JOBVR, N, A, LDA, WR, WI, VL, LDVL, VR, LDVR, 
//     $     WORK, INFO )
	FREE(pdblWork);
	if (INFO!=0)
	{
		SciError(24);
		return 0;
	}
	if(Lhs==1)
	{
		// Fill eigenvalues with computed real and imaginary parts
		int INCX = 1;
		int INCY = 1;
		C2F(dcopy)(&iCols,pdblEigenvalReal,&INCX,pdblFinalEigenvaluesReal,&INCY);
		C2F(dcopy)(&iCols,pdblEigenvalImg,&INCX,pdblFinalEigenvaluesImg,&INCY);
// DCOPY(N,DX,INCX,DY,INCY)
	} else {
		assembleComplexEigenvaluesFromDoublePointer(iRows, pdblEigenvalReal, pdblEigenvalImg, 
			pdblFinalEigenvaluesReal, pdblFinalEigenvaluesImg);

		assembleEigenvectorsSourceToTarget(iRows, pdblEigenvalImg, 
			pdblRightvectors, 
			pdblFinalEigenvectorsReal, pdblFinalEigenvectorsImg);

	}
      
	if (Lhs==1)
	{
         LhsVar(1) = 2;
	}
	else
	{

         LhsVar(1) = 3;
         LhsVar(2) = 2;
		 FREE(pdblRightvectors);
	}
	FREE(pdblEigenvalReal);
	FREE(pdblEigenvalImg);
	return 0;
}
